File: code/algo/getoffer/c++/2/window_avg.cpp
#include <vector>

class MovingAverage {
public:
  MovingAverage(int size) : size(size), sum(0) {}

  double next(int val) {
    if (q.size() == size) {
      sum -= q.front();
      q.erase(q.begin());
    }
    sum += val;
    q.push_back(val);
    return static_cast<double>(sum) / q.size();
  }

private:
  int size;
  int sum;
  std::vector<int> q;
};

File: code/algo/getoffer/go/2/window_avg.go
package main

type MovingAverage struct {
	size, sum int
	q         []int
}

func Constructor(size int) MovingAverage {
	return MovingAverage{size: size}
}

func (m *MovingAverage) Next(val int) float64 {
	if len(m.q) == m.size {
		m.sum -= m.q[0]
		m.q = m.q[1:]
	}
	m.sum += val
	m.q = append(m.q, val)
	return float64(m.sum) / float64(len(m.q))
}
